pipeline/tasks/io_tasks.py

```diff
@@ -0,0 +1,128 @@
+# limitations under the License.
+
+"""Tasks related to I/O."""
+
+import base64
+import os
+import subprocess
+import shutil
+import urllib
+import zipfile
+
+from gcloud import storage
+from oauth2client.client import GoogleCredentials
+from pipeline.tasks import task_base
+
+DEFAULT_PROJECT = 'vkit-pipeline'
+
+
+def get_credential():
+    return GoogleCredentials.get_application_default()
+
+
+class BlobUploadTask(task_base.TaskBase):
+    """A task which uploads file to Google Cloud Storage.
+
+    It requires authentication be properly configured."""
+
+    default_provides = ('bucket', 'path', 'public_url')
+
+    def execute(self,
+                bucket_name,
+                src_path,
+                dest_path,
+                project=DEFAULT_PROJECT):
+        print "Start blob upload"
+        client = storage.Client(project=project, credentials=get_credential())
+        bucket = client.get_bucket(bucket_name)
+        blob = bucket.blob(dest_path)
+        with open(src_path, 'r') as f:
+            blob.upload_from_file(f)
+        print "Uploaded to %s" % blob.public_url
+
+        return bucket_name, dest_path, blob.public_url
+
+
+class BlobDownloadTask(task_base.TaskBase):
+    """A task which downloads file to Google Cloud Storage.
+
+  It requires authentication be properly configured."""
+
+    def execute(self, bucket_name, path, output_dir, project=DEFAULT_PROJECT):
+        client = storage.Client(project=project, credentials=get_credential())
+        bucket = client.get_bucket(bucket_name)
+        blob = bucket.get_blob(path)
+        if blob:
+            filename = output_dir + '/' + path
+            if not os.path.exists(os.path.dirname(filename)):
+                try:
+                    os.makedirs(os.path.dirname(filename))
+                except:
+                    raise
+            with open(filename, "w") as f:
+                blob.download_to_file(f)
+                print 'File downloaded to %s' % f.name
+        else:
+            print 'Cannot find the output from GCS.'
+
+
+class PrepareUploadDirTask(task_base.TaskBase):
+    """Compress pipeline dir_to_upload into a file which can be uploaded to GCS.
+
+    Normally be used as the final step for pipeline job to return generated
+    content to its poster."""
+
+    def execute(self, repo_root, tarfile):
+        subprocess.call(['tar', '-C', repo_root, '-zcvf', tarfile, '.'])
+
+
+class CleanupTempDirsTask(task_base.TaskBase):
+    """Clean up all temporary directories"""
+
+    def execute(self, repo_root):
+        if os.path.exists(repo_root) and os.path.isdir(repo_root):
+            shutil.rmtree(repo_root)
+
+
+class PrepareDirsTask(task_base.TaskBase):
+
+    default_provides = ('repo_dir')
+
+    def execute(self, repo_root, files_dict={}):
+        if os.path.exists(repo_root):
+            # Do nothing if the repo_root exists. The repo_root exists if
+            # artman is running locally.
```

Comment:
Good catch. It works locally (and that was intended). I didn't document it to give myself the flexibility to remove this feature from local mode in the future.

If you specify a non-existing directory in the local mode. It will prepare the local googleapis dir and optionally copy the extra proto/yamls files into the temp dir. I found this useful for local mode.